Layout Style Guide: Add content for Lesson layout

thatblindgeye opened the pull request:
<!-- Thank you for taking the time to contribute to The Odin Project. In order to get this pull request (PR) merged in a reasonable amount of time, you must complete this entire template. -->

## Because
<!-- Summarize the purpose or reasons for this PR, e.g. what problem it solves or what benefit it provides. -->

Going through Foundations, currently some lessons include assignments with level 4 headings or post-assignment sections, neither of which are mentioned in the style guide. Short of a complete lesson rewrite/overhaul, these scenarios may be necessary at times. 

Not opposed to overhauling lessons to remove post-assignment sections either, however. Some stuff may be better suited as an additional resource or something


## This PR
<!-- A bullet point list of one or more items describing the specific changes. -->
- Adds content regarding optional assignment headings
- Adds content regarding post-assignment sections for lessons

## Issue
<!--
If this PR closes an open issue in this repo, replace the XXXXX below with the issue number, e.g. Closes #2013.

If this PR closes an open issue in another TOP repo, replace the #XXXXX with the URL of the issue, e.g. Closes https://github.com/TheOdinProject/curriculum/issues/XXXXX

If this PR does not close, but is related to another issue or PR, you can link it as above without the 'Closes' keyword, e.g. 'Related to #2013'.
-->
Closes #XXXXX

## Additional Information
<!-- Any other information about this PR, such as a link to a Discord discussion. -->


## Pull Request Requirements
<!-- Replace the whitespace between the square brackets with an 'x', e.g. [x]. After you create the PR, they will become checkboxes that you can click on. -->
-   [x] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/.github/blob/main/CONTRIBUTING.md)
-   [x] The title of this PR follows the `location of change: brief description of change` format, e.g. `Intro to HTML and CSS lesson: Fix link text`
-   [ ] The `Because` section summarizes the reason for this PR
-   [ ] The `This PR` section has a bullet point list describing the changes in this PR
-   [ ] If this PR addresses an open issue, it is linked in the `Issue` section
-   [ ] If any lesson files are included in this PR, they have been previewed with the [Markdown preview tool](https://www.theodinproject.com/lessons/preview) to ensure it is formatted correctly
-   [ ] If any lesson files are included in this PR, they follow the [Layout Style Guide](https://github.com/TheOdinProject/curriculum/blob/main/LAYOUT_STYLE_GUIDE.md)

ManonLef commented on LAYOUT_STYLE_GUIDE.md:
```diff
@@ -141,13 +151,15 @@ POST-ASSIGNMENT SECTION CONTENT.
+1. `### Assignment`: A numbered list of items that describe detailed requirements or user stories that must be followed in order to complete the project. Replace the `A REQUIREMENT/USER STORY.` with your own requirement, then add any additional numbered requirement items.
 
-3. `### Assignment`: A numbered list of items that describe detailed requirements or user stories that must be followed in order to complete the project. Replace the `A REQUIREMENT/USER STORY.` with your own requirement, then add any additional numbered requirement items.
+1. `#### CUSTOM ASSIGNMENT HEADING`: *Optional*. A heading that describes the assignment content that follows. This can be useful for breaking up an assignment into multiple lists of instructions.
```
```suggestion
1. `#### CUSTOM ASSIGNMENT HEADING`: *Optional*. A heading that describes the assignment content that follows. This can be useful for breaking up an assignment into multiple lists of instructions. If an assignment is intended to have multiple lists, each list should include a custom heading, otherwise this custom heading can be omitted.
```

Not sure if this should be added just as it is in the lesson template?

thatblindgeye replied:
True. Maybe instead of including the assignment heading text in the template we just mention inline in the "Assignment" description that headings can be added?

ManonLef replied:
That could work. Either way would be more in line with how the lesson template describes it.

ManonLef replied:
Just to be sure, the suggestion I made was for `line 151` which is in the `project layout` explanation  to be more in line with the `lesson layout` explanation. I se your made edits to the `lesson layout` description of this.

ManonLef reviewed (commented):
Hi Eric! I left some comments on the changes and wanted to note these two things as well. 
Would love to hear your thoughts on them!

- [the lesson template](https://github.com/TheOdinProject/curriculum/blob/main/templates/lesson-template.md) also needs updating after this PR is wrapped up.
- [the lesson example would also be nice to update](https://github.com/TheOdinProject/curriculum/blob/main/templates/lesson-example.md)

thatblindgeye commented on LAYOUT_STYLE_GUIDE.md:
```diff
@@ -58,11 +58,17 @@ CUSTOM SECTION CONTENT.
 
 <div class="lesson-content__panel" markdown="1">
 
+#### CUSTOM ASSIGNMENT HEADING
```
Was originally thinking about that and ended up following how other optional things were written. Does make sense to add that verbiage inline with the codeblock, though, so will make updates.

thatblindgeye commented on LAYOUT_STYLE_GUIDE.md:
```diff
@@ -58,11 +58,17 @@ CUSTOM SECTION CONTENT.
 
 </div>
 
+### POST-ASSIGNMENT SECTION HEADING
```
Yeah this was something else that was landed on I think after conversations at the time. Believe the intent is for the stuff that needs to be customized by the user to stand out more so that it doesn't as easily get forgotten and left in

thatblindgeye commented on LAYOUT_STYLE_GUIDE.md:
```diff
@@ -97,15 +103,17 @@ This section contains helpful links to related content. It isn’t required, so
    **If an assignment item does not have any instructions, remove the bulleted `AN INSTRUCTION ITEM` text from it.**
 
-5. `### Knowledge check`: A bulleted list of specific questions that a user should be able to answer on their own after reading the lesson and completing any assignment or practice. A knowledge check should only link either to a section within the lesson (either with a Heading 3 `###` or Heading 4 `####`, or by wrapping text in a `<span>` element with an `id` attribute), or a resource previously  ...
+1. `### POST-ASSIGNMENT SECTION HEADING`: *Optional*. A section that contains content that should come after the actual lesson assignment. This content should act as material that is not necessary to complete the lesson, but rather acts as optional, helpful material related to the lesson. This section will most likely not be needed for most lessons, but when it is needed simply replace the `POST- ...
```
Agreed about project post-assignments. Stuff like the Recipes project going over how to share a project from GitHub makes sense as post-assignment.

The normal lessons, though... Yeah I'm still iffy on post-assignment stuff 😆 [Git basics ](https://www.theodinproject.com/lessons/foundations-git-basics) is one example where there's a lot of post-assignment stuff. Some notes I jotted down include removing the "Cheatsheet" section, turning the "Note/warning" section into a notebox instead. The other two sections could probably just be turned into Additional Resource items if we want to find some good links.

Honestly I'd be fine with removing this update about post-assignment sections for lessons. I think a "knowledge check" and "additional resources" is honestly enough.

thatblindgeye replied:
Yeah there's definitely some things I've noticed that I think could be worked on, but is out of scope for the purposes of issues related to the layout 😭

thatblindgeye commented on LAYOUT_STYLE_GUIDE.md:
```diff
@@ -97,15 +103,17 @@ This section contains helpful links to related content. It isn’t required, so
    In order to link to a `<span>` element within the lesson, replace the value within the parenthesis with the exact `id` attribute of the `<span>` element (this will be case sensitive). For example, a `<span id="Knowledge-Check-3">` element would be linked to with `(#Knowledge-Check-3)`.
 
-6. `### Additional resources`: A bulleted list of optional resources for the user to read. Additional resources should be related to the content of the lesson in some way, without being necessary to gain an understanding of the lesson content. An additional resource should include brief text that further informs the user on why it is included or what purpose it serves.
+1. `### Additional resources`: A bulleted list of optional resources for the user to read. Additional resources should be related to the content of the lesson in some way, without being necessary to gain an understanding of the lesson content. An additional resource should include brief text that further informs the user on why it is included or what purpose it serves.
```
Yeah I can add that verbiage in. Maybe something like:

```
A bulleted list of optional resources for the user to read. Additional resources should be related to the content of the lesson in some way, without being necessary to gain an understanding of the lesson content. An additional resource should include brief text that further informs the user on why it is included or what purpose it serves, and generally should not cover the same exact content as the lesson just in a different medium. If the only answer to "What does this resource cover that the lesson doesn't?" is "I liked it" or "It's a video", it probably isn't enough to consider adding.
```

thatblindgeye commented:
@ManonLef made some edits based on your feedback, let me know what you think!

ManonLef commented on LAYOUT_STYLE_GUIDE.md:
```diff
@@ -97,15 +98,15 @@ This section contains helpful links to related content. It isn’t required, so
    In order to link to a `<span>` element within the lesson, replace the value within the parenthesis with the exact `id` attribute of the `<span>` element (this will be case sensitive). For example, a `<span id="Knowledge-Check-3">` element would be linked to with `(#Knowledge-Check-3)`.
 
-6. `### Additional resources`: A bulleted list of optional resources for the user to read. Additional resources should be related to the content of the lesson in some way, without being necessary to gain an understanding of the lesson content. An additional resource should include brief text that further informs the user on why it is included or what purpose it serves.
+1. `### Additional resources`: A bulleted list of optional resources for the user to read. Additional resources should be related to the content of the lesson in some way, without being necessary to gain an understanding of the lesson content. An additional resource should include brief text that further informs the user on why it is included or what purpose it serves, and generally should not co ...
```
```suggestion
1. `### Additional resources`: A bulleted list of optional resources for the user to read. Additional resources should be related to the content of the lesson in some way, without being necessary to gain an understanding of the lesson content. An additional resource should include brief text that further informs the user on why it is included or what purpose it serves, and generally should not cover the same exact content as the lesson but in a different medium. If the only answer to "What does this resource cover that the lesson doesn't?" is "I liked it" or "It's a video", it probably isn't enough to consider adding.
```

> same exact content as the lesson just in a different medium.

I feel this sentence needs a bit of tweaking but not sure how yet. 
Also; I think a different medium isn't a bad thing. If the lesson only covers text based resources (which is preferred), I think a recap video with the additional comment: "For those of you who'd like a video based recap of the subject presented here..." doesn't do any harm. We often see that in the lesson content itself as well.

thatblindgeye commented:
@ManonLef 

- It was intentional, agreed on that we only need a Codepen link. I figured keeping the maintainer instructions here would just require less having to go back and forth between different docs. One source of truth and all.
- Moved it to before the table of contents, though I also removed the heading.

I also added a section about Markdown styling that I kept meaning to add but kept forgetting 😆 What are your thoughts on it?

ManonLef commented:
wrong button on the review, but I'll happily oblige 
😅

ManonLef commented:
I like it @thatblindgeye ! Great addition with the markdown header too!
